src/main/java/biz/paluch/logging/gelf/intern/sender/GelfTCPSender.java

```diff
@@ -138,10 +159,35 @@ protected void write(ByteBuffer buffer) throws IOException {
                 // indicator the socket was closed
                 Closer.close(channel());
                 throw new SocketException("Cannot write buffer to channel");
+            } if (written == 0) {
+                nothingWrittenTimesInRow++;
+                if (nothingWrittenTimesInRow > writeBackoffThreshold) {
+                    if (totalSleepTimeMs > maxWriteBackoffTimeMs) {
+                        Closer.close(channel());
+                        throw new SocketException("Cannot write buffer to channel, no progress in writing");
+                    }
+                    totalSleepTimeMs += sleep(writeBackoffTimeMs);
+                }
+            } else { // written > 0
+                nothingWrittenTimesInRow = 0;
             }
         }
     }
 
+    private static long sleep(long millis) {
+        long startTime = System.nanoTime();
+        try {
+            Thread.sleep(millis);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
```

Comment:
The `sendMessage(…)` loop needs to exit after `InterruptedException` otherwise it's likely that we prevent thread shutdown. We can address this change during the merge if you want to.